[v1.18] monitor: fix cache references + overlay and L3 packets decoding support in Dissect

smagnani96 opened the pull request:
Manual backport of:

* https://github.com/cilium/cilium/pull/40308
* https://github.com/cilium/cilium/pull/40307

Doing to soften the pain for future fixes/changes to the script.

Once this PR is merged, a GitHub action will update the labels of these PRs:
```upstream-prs
 40308 40307
```

joestringer commented on pkg/monitor/datapath_debug.go:
Super minor nit comment, but the `upstream commit` references in the git logs here will change once https://github.com/cilium/cilium/pull/40308 is merged.

smagnani96 replied:
Yup, just refreshed it.
Thanks Joe 🙏🏼

smagnani96 committed:
monitor: add overlay and l3 packets support in Dissect

[ upstream commit 656f8505ea ]

[ backporter's notes: dropping TestDissect given prints are straight to
                      stdout and not to a buffer (#39957 not being backported) ]

We recently added the capability to decode Overlay packets in Monitor/Hubble.
However, this applies only to GetConnectionSummary(), which is normally
called with log level Info. In case of verbose logging, we call Dissect(),
which right not has no support for Overlay packets, printing only the
outer (underlay) layers.
Let's adjust this by aligning Dissect() to GetConnectionSummary().

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
monitor:dissect: add overlay and l3 packets to DissectSummary

[ upstream commit b17f556f19 ]

Given the latest changes to GetConnectionSummary and in the previous
commit to Dissect, let's now sync also DissectSummary.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
monitor:overlay: fix cache with old tcp state

[ upstream commit 89b9baa8e9 ]

When decoding an overlay packet, we correctly decode the packet and create
a connection info struct with all the needed bits. However, outer functions
such as getTCPInfo() are not aware whether the packet was native or overlay.
Inside, they refer `cache.tcp`, but in case of an overlay packet they should
use `cache.overlay.tcp`, which is the correct layer referring to the
inner packet.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
monitor:overlay: fix cache with old eth state

[ upstream commit 24c61cf7cc ]

Similar as in the previous commit, while printing connection info, the info
concerning the eth layer are retrieved from the cache and not the
connection structure previously created. This means that we must be aware
whether referring `cache.eth` or `cache.overlay.eth`.
This commits adjusts this. Let's test this by first decodin an ACK native
packet (also when L3 packet), then a SYN VXLAN, then an ACK GENEVE, and
finally a native SYN.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

joestringer commented on pkg/monitor/dissect.go:
```diff
@@ -342,52 +342,107 @@ func GetConnectionSummary(data []byte, opts *decodeOpts) string {
+	if !dissect {
+		fmt.Println(hex.Dump(data))
+		return
+	}
```
This commit seems to differ quite a bit from the original commit in the `Dissect()` function. Notably the upstream code has `dissectLock` involved and this commit does not. Is that lock needed here?

I would generally advocate to just ensure the backport matches the original commit as closely as possible, but in this case as I understand there are already some feature changes going on in `main` that makes this more difficult. As such, for this specific backport the approach of making this `Dissect()` implementation match the `main` branch as closely as possible seems worthwhile. That will mean we have fewer variants of this function to maintain sooner (basically when v1.17 goes out of maintenance late next year).

smagnani96 replied:
You are absolutely right, I've missed that line while solving cherry-pick conflicts.
Let me double-check those commits again 🙏🏼 

Yeah deciding not to backport PR 39957 makes it a bit harder, since almost all methods in this file were modified.

smagnani96 replied:
The cherry-picked commit you were referring to has not been adapted to preserve the old `Dissect` method while also applying the needed changes. This should now result in less code showing up as modified. Thanks again!

joestringer reviewed (commented):
I noticed another divergence from the original commit that wasn't mentioned in the commit message below, seems like it could be consequential.

joestringer reviewed (approved):
LGTM compared to upstream changes.
